File: actin-common/src/main/java/com/hartwig/actin/clinical/feed/questionnaire/QuestionnaireEntry.java
package com.hartwig.actin.clinical.feed.questionnaire;

import java.time.LocalDate;

import com.hartwig.actin.clinical.feed.FeedEntry;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class QuestionnaireEntry implements FeedEntry {

    @NotNull
    public abstract String subject();

    @NotNull
    public abstract String parentIdentifierValue();

    @NotNull
    public abstract LocalDate authoredDateTime();

    @NotNull
    public abstract String questionnaireQuestionnaireValue();

    @NotNull
    public abstract String description();

    @NotNull
    public abstract String itemText();

    @NotNull
    public abstract String itemAnswerValueValueString();

}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/questionnaire/QuestionnaireEntry.kt
package com.hartwig.actin.clinical.feed.emc.questionnaire

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.hartwig.actin.datamodel.clinical.provided.JacksonSerializable
import com.hartwig.actin.clinical.feed.emc.FeedEntry
import com.hartwig.actin.clinical.feed.emc.FeedSubjectDeserializer
import com.hartwig.actin.clinical.feed.emc.FeedValidation
import com.hartwig.actin.clinical.feed.emc.FeedValidator
import java.time.LocalDate

@JacksonSerializable
data class QuestionnaireEntry(
    @JsonProperty("subject")
    @JsonDeserialize(using = FeedSubjectDeserializer::class)
    override val subject: String,
    @JsonProperty("authored")
    val authored: LocalDate,
    @JsonProperty("description")
    val description: String,
    @JsonProperty("item_text")
    val itemText: String,
    @JsonProperty("text")
    val text: String
) : FeedEntry

class QuestionnaireEntryFeedValidator : FeedValidator<QuestionnaireEntry> {
    override fun validate(feed: QuestionnaireEntry): FeedValidation {
        return FeedValidation(QUESTIONNAIRE_DESCRIPTIONS.contains(feed.description))
    }

    companion object {
        private val QUESTIONNAIRE_DESCRIPTIONS = setOf("INT Consult", "consultation")
    }
}